File: lenet_sequential/Convolution.cpp
// #include <iostream>
// #include <vector>
#include "Convolution.h"


void convolve2D(const std::vector<std::vector<int>>& input, 
                std::vector<std::vector<int>>& output, 
                const std::vector<std::vector<int>>& kernel) {
    int kernelSize = kernel.size();
    int kCenter = kernelSize / 2;
    int rows = input.size();
    int cols = input[0].size();

    for (int i = 0; i < rows; ++i) {      
        for (int j = 0; j < cols; ++j) {  
            for (int m = 0; m < kernelSize; ++m) {   
                int mm = kernelSize - 1 - m;      
                for (int n = 0; n < kernelSize; ++n) { 
                    int nn = kernelSize - 1 - n; 
                    
                    int ii = i + (m - kCenter);
                    int jj = j + (n - kCenter);

                    if (ii >= 0 && ii < rows && jj >= 0 && jj < cols)
                        output[i][j] += input[ii][jj] * kernel[mm][nn];
                }
            }
        }
    }
}

File: python/convolution.py
from scipy import signal
import numpy as np


# Redefine 3x3 input and kernel for demonstration
input_3x3 = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
])

kernel_3x3 = np.array([
    [1, 0, -1],
    [1, 0, -1],
    [1, 0, -1]
])

# Perform convolution using 'same' mode to keep the output size equal to the input size
output_3x3 = signal.convolve2d(input_3x3, kernel_3x3, boundary='fill', mode='same')

print(output_3x3)
